Fix analyzers

tillig opened the pull request:
Associated with #3001.

## Changes

Fixes ~120 analyzer issues:

- Put types in namespaces
- Forward cancellation tokens to child operations
- Use `.Count` instead of `.Count()` for performance.
- Modifiers are not ordered.
- Use `switch` expression.
- `Dispose()` should call `GC.SuppressFinalize(this)`.
- Use pattern matching (`!(foo is object)` should be `foo is not object`).
- `default` expression can be simplified.
- Non-constant fields should not be visible outside the class.
- Collection initialization can be simplified.
- Log message templates should not change between calls (use constants, no string interpolation).
- Avoid 0-length array allocations, use `Array.Empty` instead.
- Use `Any()` instead of `Count() > 0`
- Make fields readonly where possible.
- Member doesn't access instance data and can be marked static.
- Unnecessary value assignment to variable (use discard `_` or remove the assignment if it isn't used).
- Conditional can be simplified (use `??` null coalescing op)
- Turned off the analyzers that talk about simplifying range (`[1..3]`) and indexing (`[^2]`) since those are not supported by .NET 4.6.1.

Admittedly, this is a lot for one PR so I get it if folks might want to break it up. On the other hand, there's not any real functionality change and it continues on the path of making work in VS Code waaaaay less noisy. It also saves time on reviewing, I think, given it's one PR instead of one for every commit I made (19).

tillig commented:
Hmmm. Looks like some of the range/index operators are causing trouble on Windows. I'll look at it.

cijothomas commented:
> Turns out range/index operators aren't supported in .NET 4.6.1. I reverted those changes and turned those analyzers off.

We had a similar discussion in another context - https://github.com/open-telemetry/opentelemetry-dotnet/discussions/2790#discussioncomment-2200865

tillig commented:
This is still open pending review and merge.

reyang commented on src/OpenTelemetry.Exporter.Console/ConsoleActivityExporter.cs:
```diff
@@ -71,9 +71,7 @@ public override ExportResult Export(in Batch<Activity> batch)
-                        var array = tag.Value as Array;
-
-                        if (array == null)
+                        if (tag.Value is not Array array)
```
I didn't know this before (that `is not` can be associated with a local variable), learned something new today! 😆

reyang commented on test/OpenTelemetry.Instrumentation.Grpc.Tests/Services/GreeterService.cs:
```diff
@@ -31,7 +31,7 @@ public GreeterService(ILoggerFactory loggerFactory)
         public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContext context)
         {
-            this.logger.LogInformation($"Sending hello to {request.Name}");
+            this.logger.LogInformation("Sending hello to {Name}", request.Name);
```
Nice! 👍

cijothomas replied:
🍨 for anyone who migrates to do structured logging only :)

reyang commented on test/OpenTelemetry.Tests/Logs/LogRecordTest.cs:
```diff
@@ -74,27 +75,26 @@ public void CheckCateogryNameForLog()
         }
 
         [Fact]
+        [SuppressMessage("CA2254", "CA2254", Justification = "While you shouldn't use interpolation in a log message, this test verifies things work with it anyway.")]
```
👍👍👍

cijothomas replied:
👍

CodeBlanch commented on test/OpenTelemetry.Exporter.Jaeger.Tests/JaegerExporterOptionsTests.cs:
```diff
@@ -29,6 +29,7 @@ public JaegerExporterOptionsTests()
         public void Dispose()
         {
             ClearEnvVars();
+            GC.SuppressFinalize(this);
```
Just as a FYI for these test classes which are `IDisposable` what I usually do to resolve a lot of warnings is just declare them `sealed`. Doing that relaxes a lot of the rules for proper dispose pattern because much of that pattern is there to make sure disposal works correctly in inheritance chains. Sealing class prevents those chains, which aren't particularly useful for one-off test classes anyway 😄

tillig replied:
I've done that in the past, too. Sealed works for the one where the analyzer wants the protected virtual Dispose implementation but doesn't stop the one about GC finalizer suppression.

CodeBlanch replied:
Interesting! I just tried in VS in a solution where I have analyzers running...

```csharp
		public class TestClass : IDisposable
		{
			// Generates CA1816 + CA1063
			public void Dispose()
			{
			}
		}

		public sealed class TestClass2 : IDisposable
		{
			// No warnings
			public void Dispose()
			{
			}
		}
```

Maybe it is a VS Code thing?

Anyway not a big deal just mainly curious.

cijothomas commented on test/OpenTelemetry.Tests/Resources/OtelServiceNameEnvVarDetectorTests.cs:
```diff
@@ -30,15 +30,12 @@ public OtelServiceNameEnvVarDetectorTests()
         }
 
         [Fact]
         public void OtelServiceNameEnvVar_EnvVarKey()
```
🙄

cijothomas commented on src/OpenTelemetry/Resources/Resource.cs:
```diff
@@ -72,7 +72,7 @@ public Resource Merge(Resource other)
                 foreach (var attribute in other.Attributes)
                 {
-                    if (!newAttributes.TryGetValue(attribute.Key, out var value))
+                    if (!newAttributes.TryGetValue(attribute.Key, out _))
```
`ContainsKey` instead.?

cijothomas commented:
@tillig We couldn't merge this earlier, and unfortunately now there are merge conflicts.. Could you resolve them?

tillig commented:
Merged and fixed some typos I found along the way ("Configuraion," etc. - seems like having a spell checker recommended plugin, at least for VS Code, wouldn't be a bad thing).

New warnings showed up, which I fixed as part of the merge.

There are quite a few "unused parameter" messages that are at "info" level that I think folks are ignoring but could be indicators of actual defects. For example, in `OpenTelemetry.Shims.OpenTracing.ScopeManagerShim` the [constructor takes a `Tracer` parameter](https://github.com/open-telemetry/opentelemetry-dotnet/blob/6c655698fe5f7f62593b73f754a58227a77d2159/src/OpenTelemetry.Shims.OpenTracing/ScopeManagerShim.cs#L38) but any time a `Tracer` is needed, that parameter value is ignored and [the global static `Tracer` is used instead](https://github.com/open-telemetry/opentelemetry-dotnet/blob/6c655698fe5f7f62593b73f754a58227a77d2159/src/OpenTelemetry.Shims.OpenTracing/ScopeManagerShim.cs#L54). That might be intentional, but from an API standpoint it's probably pretty unexpected. It might be worth turning some things like that up to `warning` level since it doesn't seem like folks really pay much attention to the `info` level analyzer messages. (In VS Code, they're all in one window and intermixed with info shown by default so it's harder to ignore; in VS I think info messages are turned off by default so it's pretty easy to miss them.)

cijothomas commented:
seems like having a spell checker recommended plugin, at least for VS Code, wouldn't be a bad thing -- yes agree.
